zakiyousufi/OOP-school-library
# OOP school library

### **OOP school library project**  
> _Imagine that you are the librarian of OOP University, and you need a tool to record what books are in the library and who borrows them. This simple app creates will allow you to: Add new students or teachers. Add new books. Save records of who borrowed a given book and when._ 

 **Built With**
> _Languages: `Ruby`_
> _Frameworks:` N/A`_
> _Technologies Used: `GIT`, `GITHUB`, `LINTERS`,_
> _Testes : `Unit test with Spec`_

_In this project, I have followed these requirements_ 

______________________________ Create classes `person`, `student`, and `teacher` ________________________________

- **Create class Person with the following:**
> - Instance vars: @id, @name, and @age.
> - Constructor with name, age, and parent_permission as parameter. name and parent_permission are optional and have default values of "Unknown" and true.
> - Getters for @id, @name, and @age.
> - Setters for @name and @age.
> - Private method is_of_age? that returns true if @age is greater or equal to 18 and false otherwise.
> - Public method can_use_services? that returns true if person is of age or if they have permission from parents.

- **Create class Student with the following:**

> - Inherits from Person.
> - Constructor extends parent's constructor by adding @classroom and a parameter for it.
> - Method play_hooky that returns "¯\(ツ)/¯".

- **Create class Teacher with the following:**

> - Inherits from Person.
> - Constructor extends parent's constructor by adding @specialization and a parameter for it.
> - Override can_use_services? so it always returns true.
> - Each class should be saved in a separate file.
______________________________ Decorate a class` ________________________________
- **Interface**
> - Create a class Nameable.
> > - Implement a method called `correct_name `that will raise a `NotImplementedError`.

- **Turn your Person class to Nameable**
> - Make sure that your Person class inherits from Nameable
> - Make sure that this class has a method` correct_name` implemented. It should simply return the name attribute.

- **Prepare base Decorator**
> - Make sure that it inherits from Nameable.
> - In the constructor assign a nameable object from params to an instance variable.
> - Implement the `correct_name` method that returns the result of the correct_name method of the @nameable.

- **Prepare CapitalizeDecorator and TrimmerDecorator**
> - For the CapitalizeDecorator:
> > - Create a class that inherits from the base Decorator class.
> > - Implement a method correct_name that capitalizes the output of @nameable.correct_name.
> - For the TrimmerDecorator:
> > - Create a class that inherits from the base Decorator class.
> > - Implement a method correct_name that makes sure that the output of @nameable.correct_name has a maximum of 10 characters. If it's longer it should trim the word.

______________________________ Set up Associations ________________________________
- **Create a class Classroom with the following:**
> - @label instance variable, should be initialized in the constructor.
> - Setter and getter for @label (remember about attr_accessor).

- **Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:**
> - Create the has-many side (a classroom has many students).
> - Create the belongs-to side (a student belongs to a classroom).
> - Make sure that when adding a student to a classroom it also sets the classroom for the student.
> - Make sure that when setting the classroom for a student it also adds it to the classrooms' students.

- **Create a class Book with the following:**
> @title and @author instance variables, should be initialized in the constructor.
> Setters and getters for instance variables (remember about attr_accessor).

- **Create a class Rental with the following:**
> - @date instance variable, should be initialized in the constructor.
> - Setter and getter for @date (remember about attr_accessor).

- **Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:**
> - Create the has-many side of Book and Rental (a book has many rentals).
> - Create the belongs-to side of Rental and Book (a rental belongs to a book).
> - Create the has-many side of Person and Rental (a person has many rentals).
> - Create the belongs-to side of Rental and Person (a rental belongs to a person).
> - Modify the constructor of Rental so Book and Person are set in it.

______________________________ Add basic UI ________________________________
- **Your console app should behave in the same way as it is in the example.
- Create an app.rb file that will serve as your console app entry point. It should have methods that do the following:**
> - List all books.
> - List all people.
> - Create a person (teacher or student, not a plain Person).
> - Create a book.
> - Create a rental.
> - List all rentals for a given person id.

- **In your main.rb define the entry point, this will be a method called main that is invoked at the end of your file. This method should do the following:**
> - Present the user with a list of options to perform.
> - Lets users choose an option.
> - If needed, ask for parameters for the option.
> - Have a way to quit the app.

______________________________ Preserve data ________________________________ 

- [X] Pair programming

- **On exit from your app save all the data (books, people, and rentals) into JSON files. One per type of data you are storing and named accordingly (e.g. rentals.json)**
- **On startup of your app load all data from files. Remember to handle the case when one or more of the files are missing.**

______________________________ Testing ________________________________

- [X] Pair programming

-  **Add unit tests for all the methods in all the classes:**
 > - Book
> - Person
> - Student
> - Teacher
> - Classroom
> - Rental
> - all classes that you created while implementing the Decorator design pattern
